read-fonts/src/tables/gvar.rs

```diff
@@ -0,0 +1,489 @@
+        } else {
+            (self.parent.shared_point_numbers.clone()?, var_data)
+        };
+        Some(TupleVariation {
+            axis_count: self.parent.axis_count,
+            header,
+            shared_tuples: self.parent.shared_tuples.clone(),
+            packed_deltas: PackedDeltas::new(packed_deltas),
+            point_numbers,
+        })
+    }
+}
+
+impl<'a> Iterator for TupleVariationIter<'a> {
+    type Item = TupleVariation<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.next_tuple()
+    }
+}
+
+/// A single set of tuple variation data
+#[derive(Clone)]
+pub struct TupleVariation<'a> {
+    axis_count: u16,
+    header: TupleVariationHeader<'a>,
+    shared_tuples: SharedTuples<'a>,
+    packed_deltas: PackedDeltas<'a>,
+    point_numbers: PackedPointNumbers<'a>,
+}
+
+impl<'a> TupleVariation<'a> {
+    /// Returns the 'peak' tuple for this variation
+    pub fn peak(&self) -> Tuple<'a> {
+        self.header
+            .tuple_index()
+            .tuple_records_index()
+            .and_then(|idx| self.shared_tuples.tuples().get(idx as usize).ok())
+            .or_else(|| self.header.peak_tuple())
+            .unwrap_or_default()
+    }
+
+    // transcribed from pinot/moscato
+    /// Compute the scalar for a this tuple at a given point in design space.
+    ///
+    /// Returns `None` if this tuple is not applicable at the provided coordinates.
+    pub fn compute_scalar(&self, coords: &[Fixed]) -> Option<Fixed> {
+        const ZERO: Fixed = Fixed::ZERO;
+        let mut scalar = Fixed::ONE;
+        let peak = self.peak();
+        let inter_start = self.header.intermediate_start_tuple();
+        let inter_end = self.header.intermediate_end_tuple();
+        if peak.len() != self.axis_count as usize {
+            return None;
+        }
+
+        for (i, coord) in coords.iter().copied().enumerate() {
+            let peak = peak.get(i).unwrap_or_default().to_fixed();
+            if peak == ZERO || peak == coord {
+                continue;
+            }
+
+            if coord == ZERO {
+                return None;
+            }
+
+            if let (Some(inter_start), Some(inter_end)) = (&inter_start, &inter_end) {
+                let start = inter_start.get(i).unwrap_or_default().to_fixed();
+                let end = inter_end.get(i).unwrap_or_default().to_fixed();
+                if coord <= start || coord >= end {
+                    return None;
+                }
+                if coord < peak {
+                    scalar = scalar * (coord - start) / (peak - start);
+                } else {
+                    scalar = scalar * (end - coord) / (end - peak);
+                }
+            } else {
+                if coord < peak.min(ZERO) || coord > peak.max(ZERO) {
+                    return None;
+                }
+                scalar = scalar * coord / peak;
+            }
+        }
+        Some(scalar)
+    }
```

Comment:
```suggestion
    pub fn compute_scalar(&self, coords: &[F2Dot14]) -> Option<Fixed> {
        const ZERO: Fixed = Fixed::ZERO;
        let mut scalar = Fixed::ONE;
        let peak = self.peak();
        let inter_start = self.header.intermediate_start_tuple();
        let inter_end = self.header.intermediate_end_tuple();
        if peak.len() != self.axis_count as usize {
            return None;
        }

        for i in 0..self.axis_count {
            let i = i as usize;
            let coord = coords.get(i).copied().unwrap_or_default().to_fixed();
            let peak = peak.get(i).unwrap_or_default().to_fixed();
            if peak == ZERO || peak == coord {
                continue;
            }

            if coord == ZERO {
                return None;
            }

            if let (Some(inter_start), Some(inter_end)) = (&inter_start, &inter_end) {
                let start = inter_start.get(i).unwrap_or_default().to_fixed();
                let end = inter_end.get(i).unwrap_or_default().to_fixed();
                if coord <= start || coord >= end {
                    return None;
                }
                if coord < peak {
                    scalar = scalar.mul_div(coord - start, peak - start);
                } else {
                    scalar = scalar.mul_div(end - coord, end - peak);
                }
            } else {
                if coord < peak.min(ZERO) || coord > peak.max(ZERO) {
                    return None;
                }
                scalar = scalar.mul_div(coord, peak);
            }
        }
        Some(scalar)
    }
```
Sorry for the length. This does the following:

* changes the `coords` parameter to `&[F2Dot14]`  since that is the expected type for normalized coordinates
* modifies the loop to iterate over all axes regardless of the length of coords
* replaces multiply+divide with use of the `mul_div` function which gives slightly different results but is required to match FreeType

Reply:
I have a question about this bit:

```rust
        for i in 0..self.axis_count {
            let i = i as usize;
            let coord = coords.get(i).copied().unwrap_or_default().to_fixed();
```

is this about ensuring we do something sane if `axis_count` and `coords.len()` differ? Should they ever differ? If not I would prefer just checking that they match before we enter the loop, and returning `None` if they dont?

Reply:
The reasoning here is that you can pass in a coord array that is smaller than the axis count and it just "does the right thing" in assuming all missing coords are at their default positions in design space.